### PR TITLE
Add scaling section to A2C documentation

### PR DESCRIPTION
**Pull request by Bonifatius94**
As requested by @araffin and @qgallouedec, I've added a documentation section on how to scale the A2C algorithm with SubprocVecEnv.

## Description
It's a new section in the rst documentation file for the A2C algorithm that covers scaling training steps throughput with multiple CPU cores.

## Motivation and Context
The question arose because the formerly outlined example didn't scale with CPU cores. See issue #1245.

closes https://github.com/DLR-RM/stable-baselines3/issues/1245.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist
Your CI pipeline of the PR will cover this, no worries.

### CODE REVIEW

File: docs/modules/a2c.rst [hunk 1]
```diff
@@ -76,6 +76,49 @@ Train a A2C agent on ``CartPole-v1`` using 4 environments.
...
+Scaling with CPU Cores
+----------------------
+
+The A2C algorithm allows for scaling with multiple CPU cores through parallel processing.
+To enable this feature, you'll have to override the vector environment implementation
+instantiated by ``make_vec_env()`` such that it creates a ``SubprocVecEnv``.
+
+The default ``DummyVecEnv`` implementation will only run a single thread for processing actions
+and collecting observations / rewards. With ``SubprocVecEnv``, each environment is running in its
+own thread, facilitating parallel processing for superior CPU utilization.
```
Comment (qgallouedec) [inline docs/modules/a2c.rst:88]: I would replace this by ``"(See :ref:`Vectorized Environments <vec_env>`)"``

Comment (Bonifatius94) [reply docs/modules/a2c.rst:88]: I agree, putting a link for further information on VecEnvs is a good idea. 

But I wouldn't leave out the short summary on how the different VecEnvs work. I think it fits there quite well.

But yeah, if you don't like it due to duplication for ease of maintainability, go ahead and delete it. I don't mind.

**Comment by qgallouedec** [comment]
Thank you for contributing @Bonifatius94 ! I'll review it soon.

File: docs/modules/a2c.rst (hunk 1 above)
Comment (araffin) [reply docs/modules/a2c.rst:88]: you could also link to the Notebook about multiprocessing

**Comment by Bonifatius94** [comment]
@qgallouedec I've added the cross reference you've posted. Don't know if it works though. I'm not so familiar with rst files tbh.

File: docs/modules/a2c.rst (hunk 1 above)
Comment (Bonifatius94) [reply docs/modules/a2c.rst]: @araffin Feel free to make your suggestion. I'm done for my part. You can do whatever you want with my changes 😄

**Comment by qgallouedec** [comment]
I have simplified the section dedicated to this clarification quite a bit. (I prefer to keep the documentation as synthetic as possible, although I do think that adding this clarification is justified). Actually, I think most of what you describe doesn't apply only to A2C and so I prefer to link this to the more general documentation part about it. I also corrected and simplified the python code. @araffin could you review?

**Comment by qgallouedec** [comment]
To avoid having to build the doc.

image

**Comment by Bonifatius94** [comment]
One question: Why do you think turning the GPU off is beneficial?

I know, for processing fully-connected NN layers, it doesn't really matter whether the compute is carried out on the CPU or GPU. But if that's the case, then I would assume that disabling the GPU isn't relevant at all to performance optimization :thinking: 

Maybe we should test this before merging the docs update?

**Comment by qgallouedec** [comment]
Using the example code (with 16 workers), I get 6966 FPS with GPU and 6880 without. So it seems to be equivalent.

**Comment by Bonifatius94** [comment]
Hmm yeah, I'd say leave the "disable GPU" part out? :thinking: It might be very misleading to users that are not so familiar with the SB3 codebase when the actual benefit is coming from SubprocVecEnv :thinking:

**Comment by qgallouedec** [comment]
> Using the example code (with 16 workers), I get 6966 FPS with GPU and 6880 without. So it seems to be equivalent.

@araffin can you take a look?

**Review by araffin — APPROVED** [review]
Look fine for me. The actual benefit here is hard to see as cartpole is a very simple/fast env (and the dummy vec env might be faster, see the notebook where we explore different settings). The CPU/GPU transfer cost might be minimal too.

**Comment by Bonifatius94** [comment]
> Look fine for me. The actual benefit here is hard to see as cartpole is a very simple/fast env (and the dummy vec env might be faster, see the notebook where we explore different settings). The CPU/GPU transfer cost might be minimal too.

If you have a hypothesis when and why disabling the GPU might be better, then try to set up an experiment and prove it like a scientist :wink: In case you're not familiar with optimization philosophies, have a look at this video: https://youtu.be/pgoetgxecw8?t=79 :smile:

I fear you're practicing "fake optimization" where you just say "I saw one guy on the internet do this, so it has to be that way" instead of actually measuring. If there's no real-world case backing your statement, your hypothesis most likely doesn't apply to the concrete piece of code you're trying to optimize :sweat_smile: (I know it's hard not to fall into the fake optimization trap from my own experience. But there's really no shame in getting fooled because PCs are very complicated machines :smile:)

Please consider measuring the performance and maybe even setting up some performance benchmarks such that you'll notice which changes even affect the performance. Starting to measure instead of just guessing is crucial :smile:

**Comment by qgallouedec** [comment]
I agree with you in principle. In this case, disabling the CPU will have two impacts: 
- the calculations will be done on the CPU, which, for non-image observations, is equivalent in terms of time (see https://github.com/DLR-RM/stable-baselines3/pull/1250#issuecomment-1404849705) and,
- there will be no CPU-GPU and GPU-CPU transfer. This transfer time is in fact far from negligible (see https://github.com/DLR-RM/stable-baselines3/pull/1218 where I measure it only with images). All in all, we can assume that when not using images, working on CPU is generally faster.

But assuming is not proving. I would be happy to have this performance analysis applied specifically to A2C. However, this is not a priority at the moment. In the meantime, we have to choose what you put in the doc. In my opinion, and considering the concordant clues in this specific case, it is best to put this trick in the doc, even if we discover that there is no performance gain later on.

**Comment by araffin** [comment]
> "I saw one guy on the internet do this, so it has to be that way"

my statement was more based on experience (and interaction with users: https://github.com/DLR-RM/stable-baselines3/issues/1124, https://github.com/DLR-RM/stable-baselines3/issues/413, https://github.com/DLR-RM/stable-baselines3/issues/121)

but if you want a concrete experiment:

GPU
```
python -m rl_zoo3.train --algo a2c --env BipedalWalkerHardcore-v3 --eval-freq -1 --verbose 0 -P  --vec-env subproc
```
vs
CPU only (using too many threads)
```
CUDA_VISIBLE_DEVICES= python -m rl_zoo3.train --algo a2c --env BipedalWalkerHardcore-v3 --eval-freq -1 --verbose 0 -P  --vec-env subproc 
```
vs
CPU only (using one thread only to remove communication overhead)
```
OMP_NUM_THREADS=1 CUDA_VISIBLE_DEVICES= python -m rl_zoo3.train --algo a2c --env BipedalWalkerHardcore-v3 --eval-freq -1 --verbose 0 -P --vec-env subproc
```

Gives me:

GPU:  8,500 it/s
CPU only: 7, 200 it/s
CPU only (one thread): 9,950 it/s

(I'm using a 16 core machine with NVIDIA GeForce RTX 2080 GPU)

For environment with image input, deactivating the GPU will most probably slow down or not show any benefit (I tested on Breakout).

**Pull request merged**